### ISSUE TITLE
[JENKINS-272] workspace page delivers out-dated files

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/jenkins#13599]
<p>I browse a workspace (in my case on a build slave) and view some generated<br/>
files. Then I re-build the project and try to look at the new generated files.</p>

<p>Expected: new contents<br/>
Experienced: old contents as seen when I first looked at it</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=gradopado">gradopado</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-272" target="_blank">workspace page delivers out-dated files</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Critical
<li><b>component(s)</b>: core
<li><b>resolution</b>: Cannot Reproduce
<li><b>resolved</b>: 2009-11-17T08:11:00+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 0
<li><b>imported</b>: 2025-11-24
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>I browse a workspace (in my case on a build slave) and view some generated
files. Then I re-build the project and try to look at the new generated files.</p>

<p>Expected: new contents
Experienced: old contents as seen when I first looked at it</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Platform: All, OS: All</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>gradopado:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-272?focusedId=120437&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-120437">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>very easy to proove: create a new job just with a single shell command:</p>

<p>date &gt; date.txt</p>

<p>In the worksapce view of this project, the date.txt file seems to be never updated.</p>
</pre>
</details>

</ul>
</details>

<p>very easy to proove: create a new job just with a single shell command:</p>

<p>date > date.txt</p>

<p>In the worksapce view of this project, the date.txt file seems to be never updated.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>gradopado:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-272?focusedId=120438&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-120438">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I thought it was caused by a clock difference between master and slaves, but is
is not. Time is synchronized now.</p>

<p>Found bug in 1.75, it still happens with 1.79.</p>
</pre>
</details>

</ul>
</details>

<p>I thought it was caused by a clock difference between master and slaves, but is<br/>
is not. Time is synchronized now.</p>

<p>Found bug in 1.75, it still happens with 1.79.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-272?focusedId=120439&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-120439">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I tried to reproduce the problem, but it works correctly for me.</p>

<p>What browser are you using? Can you check the HTTP headers and tell me what it
reports? For Firefox, this extension would let you see the header:
<a href="http://livehttpheaders.mozdev.org/installation.html" class="external-link" target="_blank" rel="nofollow noopener">http://livehttpheaders.mozdev.org/installation.html</a></p>

<p>What I'm interested in is the headers like 
"If-Modified-Since: Thu, 08 Feb 2007 20:13:28 PST" that the browser is sending,
and  "Last-Modified: Thu, 08 Feb 2007 20:13:28 PST" that the server will send you.</p>
</pre>
</details>

</ul>
</details>

<p>I tried to reproduce the problem, but it works correctly for me.</p>

<p>What browser are you using? Can you check the HTTP headers and tell me what it<br/>
reports? For Firefox, this extension would let you see the header:<br/>
<a href="http://livehttpheaders.mozdev.org/installation.html" class="external-link" target="_blank" rel="nofollow noopener">http://livehttpheaders.mozdev.org/installation.html</a></p>

<p>What I'm interested in is the headers like <br/>
"If-Modified-Since: Thu, 08 Feb 2007 20:13:28 PST" that the browser is sending,<br/>
and  "Last-Modified: Thu, 08 Feb 2007 20:13:28 PST" that the server will send you.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>gradopado:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-272?focusedId=120440&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-120440">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<ul>
	<li>cache cleaned, first access to the file:
Last-Modified: Fri, 09 Feb 2007 10:19:33 CET</li>
</ul>



<ul>
	<li>page reload:
If-Modified-Since: Fri, 09 Feb 2007 10:19:33 CET</li>
</ul>


<p>HTTP/1.x 304 Not Modified</p>


<ul>
	<li>"build now" @ 	09.02.2007 10:58:42</li>
	<li>access file:
(old contents, no headers sent!)</li>
</ul>


<ul>
	<li>page reload:
If-Modified-Since: Fri, 09 Feb 2007 10:19:33 CET</li>
</ul>


<p>Last-Modified: Fri, 09 Feb 2007 10:58:42 CET</p>


<ul>
	<li>"build now" @ 09.02.2007 11:00:23</li>
	<li>access file:
(old contents, no headers sent!)</li>
</ul>


<ul>
	<li>page reload:
If-Modified-Since: Fri, 09 Feb 2007 10:58:42 CET</li>
</ul>


<p>Last-Modified: Fri, 09 Feb 2007 11:00:23 CET</p>


<ul>
	<li>page reload:
If-Modified-Since: Fri, 09 Feb 2007 11:00:23 CET</li>
</ul>


<p>HTTP/1.x 304 Not Modified</p>


<p>---------------------------
Firefox just doesn't ask for a new file, at least I did not see any header. (I'm
not an expert in HTTP.) A page reload seems to help. What do you think?</p>
</pre>
</details>

</ul>
</details>

<ul>
	<li>cache cleaned, first access to the file:<br/>
Last-Modified: Fri, 09 Feb 2007 10:19:33 CET</li>
</ul>



<ul>
	<li>page reload:<br/>
If-Modified-Since: Fri, 09 Feb 2007 10:19:33 CET</li>
</ul>


<p>HTTP/1.x 304 Not Modified</p>


<ul>
	<li>"build now" @ 	09.02.2007 10:58:42</li>
	<li>access file:<br/>
(old contents, no headers sent!)</li>
</ul>


<ul>
	<li>page reload:<br/>
If-Modified-Since: Fri, 09 Feb 2007 10:19:33 CET</li>
</ul>


<p>Last-Modified: Fri, 09 Feb 2007 10:58:42 CET</p>


<ul>
	<li>"build now" @ 09.02.2007 11:00:23</li>
	<li>access file:<br/>
(old contents, no headers sent!)</li>
</ul>


<ul>
	<li>page reload:<br/>
If-Modified-Since: Fri, 09 Feb 2007 10:58:42 CET</li>
</ul>


<p>Last-Modified: Fri, 09 Feb 2007 11:00:23 CET</p>


<ul>
	<li>page reload:<br/>
If-Modified-Since: Fri, 09 Feb 2007 11:00:23 CET</li>
</ul>


<p>HTTP/1.x 304 Not Modified</p>


<p>---------------------------<br/>
Firefox just doesn't ask for a new file, at least I did not see any header. (I'm<br/>
not an expert in HTTP.) A page reload seems to help. What do you think?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-272?focusedId=120441&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-120441">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>OK, so the issue seems to be that Firefox doesn't even send a request unless you
hit F5.</p>

<p>Maybe I need to send another cache control header or something.</p>
</pre>
</details>

</ul>
</details>

<p>OK, so the issue seems to be that Firefox doesn't even send a request unless you<br/>
hit F5.</p>

<p>Maybe I need to send another cache control header or something.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-272?focusedId=120442&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-120442">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Apparently when Last-Modified header is set but no Expires header is set,
Firefox automatically guesses the effective Expires header value and won't send
GET with If-Modified-Since until that effective Expires date is reached.</p>

<p>References:
<a href="http://labs.cybozu.co.jp/blog/kazuho/archives/2006/02/utilizing_cache.php" class="external-link" target="_blank" rel="nofollow noopener">http://labs.cybozu.co.jp/blog/kazuho/archives/2006/02/utilizing_cache.php</a>
<a href="http://www.mnot.net/cache_docs/" class="external-link" target="_blank" rel="nofollow noopener">http://www.mnot.net/cache_docs/</a>
(still looking for authoritative source)</p>

<p>The solution will be to send the Expires header with the same timestamp.</p>
</pre>
</details>

</ul>
</details>

<p>Apparently when Last-Modified header is set but no Expires header is set,<br/>
Firefox automatically guesses the effective Expires header value and won't send<br/>
GET with If-Modified-Since until that effective Expires date is reached.</p>

<p>References:<br/>
<a href="http://labs.cybozu.co.jp/blog/kazuho/archives/2006/02/utilizing_cache.php" class="external-link" target="_blank" rel="nofollow noopener">http://labs.cybozu.co.jp/blog/kazuho/archives/2006/02/utilizing_cache.php</a><br/>
<a href="http://www.mnot.net/cache_docs/" class="external-link" target="_blank" rel="nofollow noopener">http://www.mnot.net/cache_docs/</a><br/>
(still looking for authoritative source)</p>

<p>The solution will be to send the Expires header with the same timestamp.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-272?focusedId=120443&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-120443">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Fixed in 1.83.</p>
</pre>
</details>

</ul>
</details>

<p>Fixed in 1.83.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>gradopado:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-272?focusedId=120444&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-120444">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I've set up a test job in hudson 1.83 which produces a file with a timestamp. It
should archive all artifacts (the file with timestamp).</p>

<p>The fix worked only once, then I still got stale files on the "latest artifacts"
page and on the workspace view also.</p>

<p>I'll check http headers if needed for your investigation.</p>
</pre>
</details>

</ul>
</details>

<p>I've set up a test job in hudson 1.83 which produces a file with a timestamp. It<br/>
should archive all artifacts (the file with timestamp).</p>

<p>The fix worked only once, then I still got stale files on the "latest artifacts"<br/>
page and on the workspace view also.</p>

<p>I'll check http headers if needed for your investigation.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-272?focusedId=120445&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-120445">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<ul>
	<li>
	<ul>
		<li>
		<ul>
			<li>Issue 390 has been marked as a duplicate of this issue. ***</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>

</pre>
</details>

</ul>
</details>

<ul>
	<li>
	<ul>
		<li>
		<ul>
			<li>Issue 390 has been marked as a duplicate of this issue. ***</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-272?focusedId=120446&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-120446">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Bumping up the priority since another report was filed for this bug.</p>
</pre>
</details>

</ul>
</details>

<p>Bumping up the priority since another report was filed for this bug.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>gradopado:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-272?focusedId=120447&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-120447">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I still have to clear the cache of my Firefox in order to get the latest build
artifacts. I mean when accessing "latestSuccessful", where the URL is always the
same but the files change.</p>

<p>Hudson version 1.152</p>
</pre>
</details>

</ul>
</details>

<p>I still have to clear the cache of my Firefox in order to get the latest build<br/>
artifacts. I mean when accessing "latestSuccessful", where the URL is always the<br/>
same but the files change.</p>

<p>Hudson version 1.152</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mindless.png" title="mindless's avatar" /> mindless:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-272?focusedId=120448&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-120448">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>As far as I can tell, the original bugfix is working correctly.  It is not clock
difference between master-slave that makes a difference, but clock difference
between where the workspace resides (in this case, slave machine) and the client
system where your browser is running.  If Hudson sends back an Expires header
for 14:00 but your system thinks it is only 13:57 then for the next couple
minutes your browser won't even send a HTTP request for that cached file.. but
once the client system's clock does reach 14:00 then another time clicking that
link does check with the server and get the latest file content.</p>

<p>Net effect: you only see old content if:
a) Your client system's clock in N minutes behind the system where the build runs
b) A build runs
c) Within N minutes of when file X is written during that build:
   1) You access file X in the workspace browser
   2) Another build runs
   3) You access file X again</p>

<p>A lot to happen in that time window.. but, if we do want to avoid this case then
the Expires header probably should be set even earlier.  Rather than matching
the Last-Modified date, make it even earlier to account for clock differences. 
Unfortunately, Stapler's serveFile method doesn't currently have any way to do
this.. any expiration value &lt;= 0 just uses the Last-Modified value.  Maybe
stapler should only do this for == 0 (which is what the javadoc says) and use
the value as a delta for both positive and negative.. then we could pass
-600000L or something.</p>

<p>Side note: only Expires header has this problem.. Last-Modified is NOT compared
to your client system's clock.. your browser remembers this value, and when it
does send another HTTP request (current time is past Expires time) then it just
sends that same value in the request, and the SERVER compares that value to the
current timestamp of file X.</p>
</pre>
</details>

</ul>
</details>

<p>As far as I can tell, the original bugfix is working correctly.  It is not clock<br/>
difference between master-slave that makes a difference, but clock difference<br/>
between where the workspace resides (in this case, slave machine) and the client<br/>
system where your browser is running.  If Hudson sends back an Expires header<br/>
for 14:00 but your system thinks it is only 13:57 then for the next couple<br/>
minutes your browser won't even send a HTTP request for that cached file.. but<br/>
once the client system's clock does reach 14:00 then another time clicking that<br/>
link does check with the server and get the latest file content.</p>

<p>Net effect: you only see old content if:<br/>
a) Your client system's clock in N minutes behind the system where the build runs<br/>
b) A build runs<br/>
c) Within N minutes of when file X is written during that build:<br/>
   1) You access file X in the workspace browser<br/>
   2) Another build runs<br/>
   3) You access file X again</p>

<p>A lot to happen in that time window.. but, if we do want to avoid this case then<br/>
the Expires header probably should be set even earlier.  Rather than matching<br/>
the Last-Modified date, make it even earlier to account for clock differences. <br/>
Unfortunately, Stapler's serveFile method doesn't currently have any way to do<br/>
this.. any expiration value <= 0 just uses the Last-Modified value.  Maybe<br/>
stapler should only do this for == 0 (which is what the javadoc says) and use<br/>
the value as a delta for both positive and negative.. then we could pass<br/>
-600000L or something.</p>

<p>Side note: only Expires header has this problem.. Last-Modified is NOT compared<br/>
to your client system's clock.. your browser remembers this value, and when it<br/>
does send another HTTP request (current time is past Expires time) then it just<br/>
sends that same value in the request, and the SERVER compares that value to the<br/>
current timestamp of file X.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/mindless.png" title="mindless's avatar" /> mindless:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-272?focusedId=120449&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-120449">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>closing old issue</p>
</pre>
</details>

</ul>
</details>

<p>closing old issue</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a href="https://github.com/jenkinsci/jenkins/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-390%22">JENKINS-390</a>]</i>
